File: src/reports.js
import ExcelJS from 'exceljs';

export async function buildReport(targetDate, rows) {
  const workbook = new ExcelJS.Workbook();
  workbook.creator = 'School Eat MAX Bot';
  workbook.created = new Date();

  const sheet = workbook.addWorksheet('Заказы', {
    views: [{ state: 'frozen', ySplit: 1 }],
    pageSetup: { fitToPage: true, fitToWidth: 1, fitToHeight: 0 },
  });

  const sortedRows = [...rows].sort(
    (left, right) =>
      left.className.localeCompare(right.className, 'ru') ||
      left.childName.localeCompare(right.childName, 'ru'),
  );
  const tableRows = sortedRows.length
    ? sortedRows.map((row) => [
        row.className,
        row.childName,
        row.breakfast ? '✓' : '—',
        row.lunch ? '✓' : '—',
      ])
    : [['—', 'Нет зарегистрированных учеников', '—', '—']];

  sheet.addTable({
    name: 'Orders',
    ref: 'A1',
    headerRow: true,
    style: { theme: 'TableStyleMedium2', showRowStripes: true },
    columns: [
      { name: 'Класс', filterButton: true },
      { name: 'ФИО', filterButton: true },
      { name: 'Завтрак', filterButton: true },
      { name: 'Обед', filterButton: true },
    ],
    rows: tableRows,
  });

  sheet.columns = [{ width: 13 }, { width: 38 }, { width: 14 }, { width: 14 }];
  sheet.getRow(1).eachCell((cell) => {
    cell.font = { color: { argb: 'FFFFFFFF' }, bold: true };
    cell.alignment = { horizontal: 'center' };
  });
  for (let rowNumber = 2; rowNumber <= sheet.rowCount; rowNumber += 1) {
    for (const columnNumber of [3, 4]) {
      const cell = sheet.getCell(rowNumber, columnNumber);
      cell.alignment = { horizontal: 'center' };
      if (cell.value === '✓') cell.font = { color: { argb: 'FF008000' }, bold: true, size: 14 };
    }
  }
  sheet.headerFooter.oddHeader = `&CЗаказы на ${targetDate.split('-').reverse().join('.')}`;

  return Buffer.from(await workbook.xlsx.writeBuffer());
}

File: app/reports.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import date
from io import BytesIO

from openpyxl import Workbook
from openpyxl.styles import Alignment, Font, PatternFill
from openpyxl.worksheet.table import Table, TableStyleInfo


@dataclass(frozen=True)
class ReportRow:
    class_name: str
    child_name: str
    breakfast: bool
    lunch: bool


def build_report(target: date, rows: list[ReportRow]) -> bytes:
    workbook = Workbook()
    sheet = workbook.active
    sheet.title = "Заказы"
    sheet.freeze_panes = "A2"
    sheet.append(["Класс", "ФИО", "Завтрак", "Обед"])

    for row in sorted(rows, key=lambda item: (item.class_name, item.child_name.casefold())):
        sheet.append(
            [
                row.class_name,
                row.child_name,
                "✓" if row.breakfast else "—",
                "✓" if row.lunch else "—",
            ]
        )

    if not rows:
        sheet.append(["—", "Нет зарегистрированных учеников", "—", "—"])

    header_fill = PatternFill("solid", fgColor="2F75B5")
    for cell in sheet[1]:
        cell.font = Font(color="FFFFFF", bold=True)
        cell.fill = header_fill
        cell.alignment = Alignment(horizontal="center")

    for row in sheet.iter_rows(min_row=2, min_col=3, max_col=4):
        for cell in row:
            cell.alignment = Alignment(horizontal="center")
            if cell.value == "✓":
                cell.font = Font(color="008000", bold=True, size=14)

    sheet.column_dimensions["A"].width = 13
    sheet.column_dimensions["B"].width = 38
    sheet.column_dimensions["C"].width = 14
    sheet.column_dimensions["D"].width = 14
    sheet.auto_filter.ref = f"A1:D{sheet.max_row}"
    sheet.sheet_properties.pageSetUpPr.fitToPage = True
    sheet.page_setup.fitToWidth = 1
    sheet.page_setup.fitToHeight = 0
    sheet.oddHeader.center.text = f"Заказы на {target:%d.%m.%Y}"

    table = Table(displayName="Orders", ref=f"A1:D{sheet.max_row}")
    table.tableStyleInfo = TableStyleInfo(
        name="TableStyleMedium2",
        showFirstColumn=False,
        showLastColumn=False,
        showRowStripes=True,
        showColumnStripes=False,
    )
    sheet.add_table(table)

    output = BytesIO()
    workbook.save(output)
    return output.getvalue()
